google_guest_agent/network/managers.go

```diff
@@ -0,0 +1,85 @@
+package network
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/run"
+	"github.com/GoogleCloudPlatform/guest-agent/metadata"
+)
+
+// ManagerService is an interface for setting up network configurations
+// using different network managing services, such as systemd-networkd and wicked.
+type ManagerService interface {
+	IsManaging(ctx context.Context, iface string) (bool, error)
+	Setup(ctx context.Context, payload []metadata.NetworkInterfaces) error
+	Rollback(ctx context.Context, payload []metadata.NetworkInterfaces) error
+}
+
+var (
+	badMAC               = []string{}
+	googleComment        = "# Added by Google Compute Engine OS Login."
+	managedKey           = "ManagedByGuestAgent=true"
+	knownNetworkManagers = []ManagerService{dhclient{}, systemdnetworkd{}}
+)
+
+func commandExists(ctx context.Context, cmd string) (bool, error) {
+	res := run.WithOutput(ctx, "command", "-v", cmd)
+	if res.ExitCode == 1 {
+		// failed to find the command.
+		return false, nil
+	} else if res.ExitCode != 0 {
+		return false, fmt.Errorf("error checking command: %v", res.StdErr)
+	} else {
+		return true, nil
+	}
+}
+
+// Gets a list of interface names.
+func interfaces(nics []metadata.NetworkInterfaces) ([]string, error) {
+	var ifaces []string
+	for _, ni := range nics {
+		iface, err := getInterfaceByMAC(ni.Mac)
+		if err != nil {
+			return nil, err
+		}
+		ifaces = append(ifaces, iface.Name)
+	}
+	return ifaces, nil
+}
+
+func getInterfaceByMAC(mac string) (net.Interface, error) {
+	hwaddr, err := net.ParseMAC(mac)
+	if err != nil {
+		return net.Interface{}, err
+	}
+
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return net.Interface{}, err
+	}
+
+	for _, iface := range interfaces {
+		if iface.HardwareAddr.String() == hwaddr.String() {
+			return iface, nil
+		}
+	}
+
+	return net.Interface{}, fmt.Errorf("no interface found with MAC %s", mac)
+}
+
+// DetectNetworkManager Detects the network manager managing the primary network interface.
+// This network manager will be used to set up secondary network interfaces.
+func DetectNetworkManager(ctx context.Context, iface string) (ManagerService, error) {
```

Comment:
can we cache this value instead of trying to detect every time?

Reply:
I think we still have to detect it every time. But caching the current one will be useful for rolling back the changes of one manager service in order to set the changes for the new one, in cases where the primary network managing service gets changed.